File: leetcode/generateMatrix/generateMatrix.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    vector<vector<int>> generateMatrix(int n) {
        // vector<vector<int>> res;
        // for(int k=0; k<n; ++k)
        //     res.push_back(vector<int>(n, 0));
        vector<vector<int>> res(n, vector<int> (n, 0));

        int i=0, j=0;
        int di=0, dj=1;

        for(int k=0; k<n*n; ++k){
            res[i][j] = k + 1;

            int it = (i+di)%n;
            if(it < 0)
                it = n + it;
            int jt = (j+dj)%n;
            if(jt < 0)
                jt = n + jt;
            if(res[it][jt]){
                int temp = di;
                di = dj;
                dj = -temp;
            }
            i += di;
            j += dj;
        }
        return res;    
    }
};


int main(){
    Solution solver;
    vector<vector<int>> vec = solver.generateMatrix(3);
    for(auto& x: vec){
        for(auto y: x){
            cout << y << " ";
        }
        cout << endl;
    }
    // cout << 2 % 3 << endl;
    // vector<int> vec{1,2,3};
    // cout << vec[0] << endl;
    // cout << vec[-1] << endl;
    
}

File: leetcode/generateMatrix/generateMatrix.py
# Given a positive integer n, generate a square matrix filled with elements from 1 to n^2 in spiral order.

# Example:

# Input: 3
# Output:
# [
#  [ 1, 2, 3 ],
#  [ 8, 9, 4 ],
#  [ 7, 6, 5 ]
# ]

# https://leetcode.com/problems/spiral-matrix-ii/discuss/22282/4-9-lines-Python-solutions
class Solution:
    def generateMatrix(self, n):
        """
        :type n: int
        :rtype: List[List[int]]
        """
        A = [[0] * n for _ in range(n)]
        i, j, di, dj = 0, 0, 0, 1
        for k in range(n*n):
            A[i][j] = k + 1
            # print((i+di)%n, (j+dj)%n)
            if A[(i+di)%n][(j+dj)%n]:
                di, dj = dj, -di
            i += di
            j += dj
        return A


if __name__ == '__main__':
    solver = Solution()
    print(solver.generateMatrix(3))





        